hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java

```diff
@@ -633,6 +636,14 @@ protected ScrollableResultsImplementor doScroll(ScrollMode scrollMode) {
 	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 	// Select query plan
 
+	@Override
+	public boolean isQueryPlanCacheable() {
+		return CRITERIA_HQL_STRING.equals( hql )
+				// For criteria queries, query plan caching requires an explicit opt-in
+				? getQueryOptions().getQueryPlanCachingEnabled() == Boolean.TRUE
```

Comment:
what about a warn about the risk of enabling it for non immutable criteria.

Reply:
Can you elaborate a bit? Why would we do that? Do you mean the warning is appropriate because the caching of such query plans could trash the cache? To me, it kind of feels unnecessary to emit a warning, because people will have to opt-in to this anyway. I can write a warning into the Javadoc or explain how the cache trashing could happen in the `@implNote` of the setting.

Reply:
explaining this in the the javadoc is perfect